### PR TITLE
New design : implementation independent API

### PR DESCRIPTION
**Pull request by blegat**
Rewrite of the code, it is now an implementation independent API. The implementation has been moved to [DynamicPolynomials](https://github.com/blegat/DynamicPolynomials.jl). The API is inspired from both [DynamicPolynomials](https://github.com/blegat/DynamicPolynomials.jl) and [TypedPolynomials](https://github.com/rdeits/TypedPolynomials.jl).
There is a start of documentation of the API.
Still to do:
- [x] substitution
- [x] make TypedPolynomials pass the tests
- [x] run the tests for both TypedPolynomials and DynamicPolynomials on `runtests.jl` of this package

@rdeits What would you think of saying that `variables` gives the variables of nonzero exponent and `vars` gives all the variables of the type ?

### CODE REVIEW

File: src/diff.jl [hunk 1]
```diff
@@ -1,46 +1,30 @@
...
-            Z[j][i] -= 1
-        end
-        Polynomial(a, MonomialVector(vars(p), Z))
-    end
-end
+_diff_promote_op(::Type{TT}, ::Type{<:AbstractVariable}) where {T, TT<:AbstractTermLike{T}} = changecoefficienttype(TT, Base.promote_op(*, T, Int))
+differentiate(t::AbstractTermLike, v::AbstractVariable) = coefficient(t) * differentiate(monomial(t), v)
+
+_diff_promote_op(::Type{PT}, ::Type{<:AbstractVariable}) where {T, PT<:APL{T}} = polynomialtype(PT, Base.promote_op(*, T, Int))
+differentiate(p::APL, v::AbstractVariable) = polynomial(differentiate.(terms(p), v))
```
Comment (saschatimme) [inline src/diff.jl]: This leads to zeros for constants, i.e.
```julia
julia> MP.differentiate(2x+3, x)
0 + 2
```
But I can also fix this after the PR got merged (just don't to make a PR on the PR^^)

Comment (blegat) [inline src/diff.jl]: It is the responsability of `polynomial` to sort terms and remove zeros but currently `TypedPolynomials` does not remove zero terms, this should be fixed indeed :)

Comment (saschatimme) [inline src/diff.jl]: Good to know, will keep that on my list and fix it later :) (Unless you fix it before everything gets merged)

**Comment by rdeits** [comment]
I was able to run the tests on this branch, but happened to run into an interesting Julia issue once I added Reexport: https://github.com/JuliaLang/julia/issues/23200

**Comment by saschatimme** [comment]
@rdeits I ran into the same Julia issur (even without reexport). It was not deterministic, but I think after I removed ˋRevise.jlˋ from my ˋjuliarc.jlˋ it didn’t occur again.

**Comment by blegat** [comment]
I also keep getting hit by https://github.com/JuliaLang/julia/issues/22355. When this happens, I use Julia v0.7. Once I do a few changes this usually disappears but then it appears a few changes later :-P They seem to say that a fix is being backported to the next bug fix release of Julia v0.6.

**Comment by blegat** [comment]
We may strip out `matpoly.jl` from MultivariatePolynomials and put it in SumOfSquares.jl. Making this package lighter might make these failures less frequent.

**Comment by saschatimme** [comment]
`algebraicset.jl` also? That seems quite self contained.

**Comment by blegat** [comment]
Both TypedPolynomials and DynamicPolynomials are passing all the tests.
@rdeits Ok to merge this PR ?
I have removed MatPolynomial, SOSPolynomials and algebraic set stuff. The two first will be moved to SumOfSquares while the last will be moved to SemialgebraicSets.jl unless someone has a better name.

**Pull request merged**